interfaces/apparmor/apparmor.go

```diff
@@ -64,33 +64,70 @@ const (
 	skipReadCache aaParserFlags = 1 << iota
 )
 
-func loadProfile(fname, cacheDir string, flags aaParserFlags) error {
+func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
+	if len(fnames) == 0 {
+		return nil
+	}
+
 	// Use no-expr-simplify since expr-simplify is actually slower on armhf (LP: #1383858)
-	args := []string{"--replace", "--write-cache", "-O", "no-expr-simplify", fmt.Sprintf("--cache-loc=%s", cacheDir)}
+	args := []string{"--replace", "--write-cache", "-O", "no-expr-simplify",
+		fmt.Sprintf("--cache-loc=%s", cacheDir)}
 	if flags&skipReadCache != 0 {
 		args = append(args, "--skip-read-cache")
 	}
 	if !osutil.GetenvBool("SNAPD_DEBUG") {
 		args = append(args, "--quiet")
 	}
-	args = append(args, fname)
+	args = append(args, fnames...)
 
 	output, err := exec.Command("apparmor_parser", args...).CombinedOutput()
 	if err != nil {
-		return fmt.Errorf("cannot load apparmor profile: %s\napparmor_parser output:\n%s", err, string(output))
+		return fmt.Errorf("cannot load apparmor profiles: %s\napparmor_parser output:\n%s", err, string(output))
+	}
+	return nil
+}
+
+func reloadProfiles(profiles []string, profileDir, cacheDir string) error {
+	files := []string{}
+	for _, profile := range profiles {
+		files = append(files, filepath.Join(profileDir, profile))
+	}
+	err := loadProfiles(files, cacheDir, 0)
+	if err != nil {
+		return fmt.Errorf("cannot load apparmor profiles %q: %s", files, err)
+	}
+	return nil
+}
+
+func reloadChangedProfiles(profiles []string, profileDir, cacheDir string) error {
+	files := []string{}
+	for _, profile := range profiles {
+		files = append(files, filepath.Join(profileDir, profile))
+	}
+	err := loadProfiles(files, cacheDir, skipReadCache)
```

Comment:
Similar case here: this function looks exactly the same as the prior function, but with skipReadCache.

Can we drop these two helpers, and just call the real underlying function with the actual flag at the call sites? The common logic may be moved into the underlying function, or to a completely different helper function that just deals with adding the prefix to the path.

This way we'd end up with a single function instead of three, and have more readable logic in the sense we know what is being done at the call site (I'd prefer knowing the call will skipReadCache than wondering what reloadChangedProfiles actually does, and be surprised that it really doesn't).

Reply:
I agree, I'll look into this. The functions have both confusing names and some redundant existence now.